### DISCUSSION TITLE
Weird behavior using @entangle on Laravel model

### DISCUSSION THREAD

**Discussion by n3b0r** [livewire/livewire#1952] [1. Help] [answered]
### Description
I'm experiencing a weird behavior using @entangle on Laravel model.

Seems to be a gap between Alpine change and model update. The weird thing is that it's only happening when I try to update model properties(i.e. user.active) but doesn't when I try to update Livewire properties.

I'm going to share my code trying to explain the issue.

IMO @entangle seems to lost the model reference or doesn't fire the change event from Alpine to Livewire by an unknown problem on the first shot.

### Exact steps to reproduce

Create two Livewire properties, one Laravel model one standalone. 

The standalone property it's reactive and works as expected. The laravel model property is not reactive to the first change. No errors are shown on the browser console.

If I switch to a native HTML component like a checkbox, everything works as expected. 

### Stripped-down, copy-pastable code snippets

**File: app\Models\User.php**

```
namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Foundation\Auth\User as Authenticatable;

class User extends Authenticatable
{
    use HasFactory;
    
    protected $fillable = [
       'active'
    ];

    protected $casts = [
        'active' => 'boolean'
    ];
}
```

**File: app\Http\Livewire\Users.php**

```
namespace App\Http\Livewire;

use App\Models\User;
use Illuminate\Foundation\Auth\Access\AuthorizesRequests;
use Livewire\Component;

class Users extends Component
{
    use AuthorizesRequests;
    
    public User $user;
    public $active = false;
    
    public function rules()
    {
        return [
        'user.active' => 'required|boolean',
        'active' => 'required|boolean',
        ];
    }
    public function mount()
    {
        $this->user = User::make(['active' => false]);
    }
}
```

**File: resources\views\livewire\users.blade.php**

```
<div>
    <x-input.group for="user.active" label="Active user?" :error="$errors->first('user.active')" toggle>
        <x-input.toggle wire:model="user.active" color="red" />
    </x-input.group>
    <x-input.group for="active" label="Active user?" :error="$errors->first('active')" toggle>
        <x-input.toggle wire:model="active" color="red" />
    </x-input.group>
</div>
```

**File: resources\views\components\input\toggle.blade.php**

```
@props([
	'color' => 'green',
	'background' => [
		'blue' => 'bg-blue-900',
		'light-blue' => 'bg-blue-500',
		'green' => 'bg-green-600',
		'red' => 'bg-red-600',
		'orange' => 'bg-orange-400',
		'yellow' => 'bg-yellow-300',
		'gray' => 'bg-gray-700'
	],
])
<div
	x-data="{ on: @entangle($attributes->wire('model')) }"
>
	<span
		x-on:click="on = !on"
		x-on:keydown.space.prevent="on = !on"
		x-bind:class="{ 'bg-gray-300': !on, '{{ $background[$color] }}': on }"
		class="relative inline-flex flex-shrink-0 h-6 w-11 border-2 border-transparent rounded-full cursor-pointer transition-colors ease-in-out duration-200 focus:outline-none bg-gray-200">
		<span x-bind:class="{ 'translate-x-5': on, 'translate-x-0': !on }" class="translate-x-0 relative inline-block h-5 w-5 rounded-full bg-white shadow transform transition ease-in-out duration-200">
			<span x-bind:class="{ 'opacity-0 ease-out duration-100': on, 'opacity-100 ease-in duration-200': !on }" class="opacity-100 ease-in duration-200 absolute inset-0 h-full w-full flex items-center justify-center transition-opacity">
				<svg class="h-3 w-3 text-gray-400" fill="none" viewBox="0 0 12 12">
					<path d="M4 8l2-2m0 0l2-2M6 6L4 4m2 2l2 2" stroke="currentColor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"></path>
				</svg>
			</span>
			<span x-bind:class="{ 'opacity-100 ease-in duration-200': on, 'opacity-0 ease-out duration-100': !on }" class="opacity-0 ease-out duration-100 absolute inset-0 h-full w-full flex items-center justify-center transition-opacity">
				<svg class="h-3 w-3 text-indigo-600" fill="currentColor" viewBox="0 0 12 12">
					<path d="M3.707 5.293a1 1 0 00-1.414 1.414l1.414-1.414zM5 8l-.707.707a1 1 0 001.414 0L5 8zm4.707-3.293a1 1 0 00-1.414-1.414l1.414 1.414zm-7.414 2l2 2 1.414-1.414-2-2-1.414 1.414zm3.414 2l4-4-1.414-1.414-4 4 1.414 1.414z"></path>
				</svg>
			</span>
		</span>
	</span>
</div>
```

**Context**
- Livewire version: 2.3.1
- Laravel version: 8.12.3
- Alpine version: 2.7.3
- Browser: Chrome

**Comment by joshhanley** [comment]
@vgsystems I have copied your code down into a local test and it all seems to be working fine for me, see screenshot. 

All I added was outputting the Livewire results to the screen just to confirm it is actually running correctly, and remove the surrounding input group in users.blade.php as I didn't have that.

I'm testing on livewire 2.3.1 and alpine 2.7.3.

Can you double check you don't have something else going on?

@vgsystems or @danharrin can you move this to discussions for now until we can confirm if this a bug in livewire?

![recording (1)](https://user-images.githubusercontent.com/882837/97931899-c6235f80-1dc2-11eb-9ad3-56262fee8075.gif)

**Comment by n3b0r** [comment]
Hi @joshhanley, thanks for taking the time to review the problem. As far as I experienced the same issue in other models, I'm going to isolate the code and publish a full project in a Github repo to check the issue. Give me some hours to publishing it.

**Reply by n3b0r** [reply]
Hi @joshhanley here is a example repo [https://github.com/vgsystems/entangle-issue](https://github.com/vgsystems/entangle-issue)

Please, let me know if you can reproduce the issue.

Thanks for your support,

![out](https://user-images.githubusercontent.com/7352160/98046056-c6a31100-1e29-11eb-9691-75a2f717370e.gif)

**Comment by joshhanley** [comment]
Ok yeah I can see the issue now, it is only when the checkbox component (which has Alpine initialised) is running inside another instance of Alpine (ie that is initialised at the page level).

I believe it is related to #1761. I'm confirming a few things over there to if it is the same issue, and if it is I will see if I can write a test that replicates this issue.

**Reply by n3b0r** [reply]
Now that you see the issue, I added an extra hint/commit to the repo. If I use the toggle with a Livewire property instead of the Laravel model everything works properly and this is what it weird me.

Please check the new commit to see it.

**Reply by n3b0r** [reply]
![out](https://user-images.githubusercontent.com/7352160/98082698-75bb0900-1e79-11eb-83c0-18413e69cdfa.gif)

**Reply by joshhanley** [reply]
Yep the reason that works is because by toggling a livewire property directly, it's sending a dedicated request back to the server (the way it was done before \@entangle).

The issue with the first example is Livewire isn't watching the Alpine changes correctly the first couple of times probably due to a bug with \@entangle.

I'll see if I can simply the example and replicated it.

**Reply by n3b0r** [reply]
Thanks!

**Reply by joshhanley** [reply]
I believe I have found the issue, have a look at my comment here https://github.com/livewire/livewire/issues/1761#issuecomment-722066719

Let me know if that is the same for you.

**Reply by n3b0r** [reply]
That's it, Josh! Thanks for that, any plan to solve it?

**Reply by joshhanley** [reply]
Yep! I am currently writing a test that clearly reproduces that matter, so it's easier to fix.

**Reply by n3b0r** [reply]
Hi @joshhanley is this solved in today's release? 

Thanks!

**Reply by joshhanley** [reply]
Probably not, I current have an open PR for another test that I need to be working before I can finalise this one. But if you get a chance to test the new release, please let us know if it's still happening.

**Reply by n3b0r** [reply]
I will during the week Josh.

**Reply by n3b0r** [reply]
Hi, @joshhanley I tested it with the last Livewire version and I can confirm that it's not fixed yet and I think that it works worst than before.

I updated the demo repo.

**Reply by joshhanley** [reply]
Yep there is a bug in the new version which needs to be fixed before we can this issue can be fixed.

**Comment by joshhanley** [comment, marked as answer]
@vgsystems there has been some fixes to entangle done! Can you try using 2.3.5 and see if your issue has been fixed?

**Reply by n3b0r** [reply]
Hey, @joshhanley now it's working like a champ! Thank you so much for following the issue and cooperate to solve it ;)

**Reply by joshhanley** [reply]
Great! No worries :)
